File: Android/shiba/src/main/java/moe/tlaster/shiba/INotifyPropertyChanged.kt
package moe.tlaster.shiba

interface INotifyPropertyChanged {
    var propertyChanged: Event<String>
}

abstract class BaseNotifyObject : INotifyPropertyChanged {
    override var propertyChanged: Event<String> = Event()

    fun notifyPropertyChanged(propertyName: String) {
        propertyChanged.invoke(this, propertyName)
    }
}

class Event<T> {
    private val listeners = ArrayList<(sender: Any, arg: T) -> Unit>()
    fun invoke(sender: Any, arg: T) {
        listeners.forEach { it.invoke(sender, arg) }
    }

    operator fun plusAssign(propertyChanged: (Any, T) -> Unit) {
        listeners.add(propertyChanged)
    }

    operator fun minusAssign(propertyChanged: (Any, T) -> Unit) {
        listeners.remove(propertyChanged)
    }

    fun clear() {
        listeners.clear()
    }
}

@Target(AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER, AnnotationTarget.FUNCTION)
@Retention(AnnotationRetention.RUNTIME)
annotation class Binding(val name: String)

File: iOS/Shiba/Shiba/INotifyPropertyChanged.swift
//
// Created by 高垣朝陽 on 2018/8/16.
// Copyright (c) 2018 Tlaster. All rights reserved.
//

import Foundation

//public class PropertyChangedHandler {
//    private var listeners: [(_ sender: Any, _ propertyName: String) -> Void] = []
//
//    static func +=(handler: PropertyChangedHandler, propertyChanged: (Any, String) -> Void) {
//        handler.listeners += propertyChanged;
//    }
//
//    static func -=(handler: PropertyChangedHandler, propertyChanged: (Any, String) -> Void) {
//        handler.listeners -= propertyChanged
//    }
//
//    func clear() {
//        self.listeners.removeAll()
//    }
//}

public protocol INotifyPropertyChanged {
    var propertyChanged: [(_ sender: Any, _ propertyName: String) -> Void] { get set }
}

public class NotifyObject: INotifyPropertyChanged {
    public var propertyChanged: [(_ sender: Any, _ propertyName: String) -> Void] = []
}
